fcrepo-persistence-api/src/main/java/org/fcrepo/persistence/api/PersistentStorageSession.java

```diff
@@ -86,6 +88,39 @@ public FedoraResource read(final String identifier) throws PersistentItemNotFoun
      */
     public FedoraResource read(final String identifier, final Instant version) throws PersistentItemNotFoundException;
 
+    /**
+     * Returns the types for the selected resource.
+     *
+     * @param identifier identifier of the resource
+     * @return list of types
+     * @throws PersistentItemNotFoundException If the identifier doesn't exist.
+     */
+    public List<String> getTypes(final String identifier) throws PersistentItemNotFoundException;
+
+    /**
+     * Read the client managed triples for the provided resource, and store them to that resource.
+     *
+     * @param resource the resource.
+     * @throws PersistentItemNotFoundException If the identifier doesn't exist.
+     */
+    public void readTriples(final FedoraResource resource) throws PersistentItemNotFoundException;
```

Comment:
I'm feeling a little uneasy with the pattern being introduced here.   I'm open to being persuaded otherwise but it seems cleaner to have separate methods that describe the various use cases we need for reading out FedoraResource or optionally a variable set of parameters in the read() method for specifying which _representation_ or _view_ of the resource is desired.   My concern (which may be unfounded) is that we are starting to blur the ideas of FedoraResource  as a conceptual entity and the notion of it as a  _view_ of it such that at any particular place in the code we may find it difficult to reason about a particular instance of FedoraResource's state.  I think this warrants a face-to-face team discussion.

Reply:
Would separate methods for the various use cases mean something like this? (names meant to be descriptive, not realistic):
`readWithTriplesAndServerManagedProperties(id)`
`readWithTriples(id)` (for Prefer omitting SMTs)
`readWithServerManagerProperties(id)`
`readWithBinaryContentAndServerManagedProperties(id)`
`readWithHeaders(id)` (for HEAD request)

or the second idea with 
`read(id, TRIPLES, SMTS)`
`read(id, TRIPLES)`
`read(id, SMTS)`
`read(id, BINARY, SMTS)`
`read(id)` (headers automatic in all reads)

I guess it seemed simpler to me to lazy load the parts of the resource that were necessary when they were requested, versus figuring out all the needed data up front. It would also accommodate multiple reads of any of the streams without needing to make a new FedoraResource. For a HEAD request ideally it wouldn't need to load any of the RDF files, so `readHeaders()` would probably defer to an index.

Worth mentioning that this PR doesn't address how headers would be loaded without populating the SMT rdf stream. So there might need to be something like `readHeaders()` added in here as well, or some other client (either against a index or the Simple Search) for getting individual properties of a resource (which might be nice to have regardless).

We can talk about it next week

Reply:
I prefer the second suggestion assuming that  you would be returning the populated FedoraResource rather than passing in an unpopulated instance.

FedoraResource read(id, FedoraResourceOption... option);

Reply:
Also I wonder if the persistentStorageSession is really what we want.  Something like this might be a more clear for the CUD methods: 

Also I'm wondering if we should consider an alternative interface for the CUD methods: 

CreateResourceRequest req = new CreateResourceRequest(id, incomingtriples, headers); 

CreateResourceResponse resp =  ps.create(req);

I'm not sure that returning the FedoraResource on create() is a good idea since FedoraResource can be in any number of different states depending on what "view" of it is required. Perhaps it is better to let the client specify exactly what he/she wants on a read (as you did above).

Reply:
In my most recent commit, I shifted things a little:
https://github.com/fcrepo4/fcrepo4/pull/1557/commits/5bc950ccb69a05d08b032086ca17f3b5e6c7d7bc
Now when a resource is read it would always start with all of its header information. The various streamed content (triples, server managed triples, binary content) would be loaded on demand. For a newly created or updated resource, `setTriples` would have been called, and the client provided rdf stream would be returned by the FedoraResource instead of pulling it from the persistence layer.